Mark move constructors and move assignment operators as noexcept

JoeLoser opened the pull request:
**Problem:**
It is generally recommended to mark move constructors and move-assignment operators as `noexcept`. 

**Solution:**
Mark user-defined move constructors and move assignment operations as `noexcept`. This will help usage with STL containers to enable using move operations (via vendor implementation detail of `move_if_noexcept`) to generally improve performance.

**Questions:**

1. Do people like putting `noexcept` when ` = defaulting` a user-defined-type? I suspect people would prefer to *not* have it as it is extra characters and noise for no perceived benefit. Currently, I *have* added it, but I'm happy to revert in favor of the fewer-character option.

**Reviewers:**
@HowardHinnant @seelabs

ripplelabs-jenkins commented:
## Jenkins Build Summary

Built from [this commit](https://github.com/ripple/rippled/commit/93a53562e1780d79c7763f3b0313ed6bc0db2903)

Built at __20180521 - 02:00:54__

### Test Results

Build Type | Log | Result | Status
---------- | --- | ------ | ------
Author Check | null | JoeLoser is not a collaborator! | FAIL :red_circle:

HowardHinnant commented:
> Do people like putting noexcept when = defaulting a user-defined-type? I suspect people would prefer to not have it as it is extra characters and noise for no perceived benefit. Currently, I have added it, but I'm happy to revert in favor of the fewer-character option.

We don't actually have a policy on this.  I think it is fine.  If you put the `noexcept` on and the special member is actually not `noexcept`, then one gets a compile-time error.  So putting it on is a good way to assure the reader that "yes, it really is noexcept".  Another technique I like to use is to assert it:
```c++
static_assert(std::is_nothrow_move_constructible<X>::value, "");
```

HowardHinnant commented:
As an overall policy, I do not believe we should make anything `noexcept` that might throw an exception, and thus call `terminate()`.  Subsequently, the function in question calls anything that is not itself marked `noexcept`, then we have a problem, whether or not said called function actually throws or not.  We have no compile-time assurance that `terminate()` won't be called.

There's a few spots marked `noexcept` that will force the compiler to insert calls to `terminate()`, and thus I think we have to remove the `noexcept`.  I'll call out each one individually in the diff.

HowardHinnant commented on src/ripple/beast/container/detail/aged_ordered_container.h:
```diff
@@ -328,7 +328,7 @@ class aged_ordered_container
         }
 
-        config_t (config_t&& other)
+        config_t (config_t&& other) noexcept
```
I believe this may call `terminate`, and I was unable to fix it by making `KeyValueCompare` `noexcept`, because that looked like it will call `terminate`.  Thus everything in this header fails without this foundation.

`aged_ordered_container` could probably be made `noexcept`, but not this easily.

HowardHinnant replied:
Hold off on this.  I think I see a way to make this work...

HowardHinnant replied:
<sigh>  I can make `config_t` work.  But `cont_type` and `chronological.list` are boost::intrusive types and they are not nothrow_move_constructible and nothrow_move_assignable respectively.  That means the ordered_container move constructor can throw.

JoeLoser replied:
@HowardHinnant I definitely overlooked the fact that if we can not prove that the whole call-def-chain is `noexcept` for all inputs, we can't mark the top function in the chain as `noexcept`. You're exactly right about the worrisome calls to `std::terminate`.

In this particular case, `boost::intrusive` types are the showstopper as you pointed out. I'm not even sure there is value in making `config_t` `noexcept-correct`, so I may just revert.

HowardHinnant commented on src/ripple/beast/core/SharedPtr.h:
```diff
@@ -176,7 +176,7 @@ class SharedPtr
     /** @{ */
 #if BEAST_SHAREDPTR_PROVIDE_COMPILER_WORKAROUNDS
-    SharedPtr& operator= (SharedPtr && sp)
+    SharedPtr& operator= (SharedPtr && sp) noexcept
```
This needs `noexcept` on `swap`.

JoeLoser replied:
Right. And in making `swap` `noexcept`, we'll have to make several other functions `noexcept`. You eventually get to a point where we're imposing potentially new requirements on the type `U` for `acquire` and the type `T` for `release`. Namely, the functions `incReferenceCount` or `decReferenceCount` must be `noexcept`.

So, to continue down this rabbit hole: we need to analyze the usages in Ripple using `beast::SharedPtr`. One such example uses `Source` which inherits from `beast::SharedObject`. We would need to make its `decReferenceCount` `noexcept` which in turn means making `destroy` `noexcept`. I'd have to look at the other uses of `SharedPtr` to verify their ref count methods are `noexcept`.

It makes me wonder though: is this worth doing? It seems fragile as if someone uses `SharedPtr` with a new type and doesn't mark their ref count methods `noexcept`, I don't see an easy way to protect against that in the future.

HowardHinnant replied:
I was thinking (the member) `swap` just swaps pointers, side stepping that rabbit hole.  I missed the `assign` which takes us into that rabbit hole.  I know it's doable, because `std::shared_ptr` does it.  We should probably follow that model.  Otoh, perhaps the right thing to do is to ditch trying to fix `SharedPtr` and instead replace it with `std::shared_ptr` in a separate P/R instead.

JoeLoser replied:
I'm with you on the latter idea. I think the right thing to do here is to just replace usages with `std::shared_ptr` and remove this dependency.  I'll create a GitHub issue for that and likely work on it next week after the holiday.

JoeLoser replied:
Ended up just doing this now -- PR is https://github.com/ripple/rippled/pull/2557

HowardHinnant commented on src/ripple/core/ClosureCounter.h:
```diff
@@ -93,7 +93,7 @@ class ClosureCounter
         }
 
-        Wrapper (Wrapper&& rhs)
+        Wrapper (Wrapper&& rhs) noexcept
```
This doesn't work because `closure_` is not always nothrow_move_constructible.

JoeLoser replied:
Good catch. We could make it conditionally noexcept though.

```
Wrapper (Wrapper&& rhs) noexcept(std::is_nothrow_move_constructible<Closure>::value)
```

HowardHinnant commented on src/ripple/ledger/CashDiff.h:
```diff
@@ -62,7 +62,7 @@ class CashDiff
     CashDiff() = delete;
     CashDiff (CashDiff const&) = delete;
-    CashDiff (CashDiff&& other);
+    CashDiff (CashDiff&& other) noexcept;
```
We could `noexcept = default` this one while we're here.

JoeLoser replied:
Right. I went ahead and default the dtor too while in the neighborhood.

JoeLoser replied:
@HowardHinnant I don't think we actually want to `noexcept = default` it here. 

My reasoning: to default it would require exposing the complete type if `Impl` when parsing the `CashDiff` type. This is because the default move ctor would require knowing the full type of the `std::unique_ptr<Impl>`.  Right now we forward declare `Impl` so we have an incomplete type. I suspect this is also the reason why the destructor (though trivial) is defined out of line -- to avoid having the complete type.

Thoughts?

HowardHinnant replied:
Good thing you're on the ball! Sounds completely right. 👍

HowardHinnant commented on src/ripple/ledger/ReadView.h:
```diff
@@ -223,7 +223,7 @@ class ReadView
     }
 
-    ReadView (ReadView&& other)
+    ReadView (ReadView&& other) noexcept
```
Need to make these statements compile:
```c++
        static_assert(std::is_nothrow_constructible<decltype(sles), ReadView&>{}, "");
        static_assert(std::is_nothrow_constructible<decltype(txs), ReadView&>{}, "");
```

JoeLoser replied:
Will work on that. Thanks! I'm also in progress of adding the usual `static_assert` checks like you mentioned earlier. For most user-defined-types, in other codebases, I usually just write `static_asserts` for all ctor, dtor, copy, copy assign, move, and move-assign (with noexcept) just to be overly explicit. Happy to adopt that same philosophy in this codebase.

HowardHinnant replied:
To be honest, we haven't been adopting this practice universally.  But I wouldn't object to it in test code.  And the occasional place like this where you need to make sure a `= default` is doing the right thing isn't a bad idea either.

JoeLoser replied:
Fair enough. I'm in the (slow) progress of being a lot more careful with this PR.  I'm making small commits and marking them with the usual `[FOLD]` so you can follow the foot-trail I leave behind. I'll ping you when I think it's in a good reviewable state again. Until then, I'll probably just close the PR and re-open when it's ready.

HowardHinnant replied:
Thanks much for you work on this.  I think this will be a valuable addition.

JoeLoser commented:
@HowardHinnant Okay, I think we're in much better shape with this PR now, so it is reopened now. Thanks for pointing out that I don't get the guarantee of `noexcept-correctness` without calls to `std::terminate` for free unless I force an instantiation in a context that requires that property (such as `static_assert`). I've gone ahead and added tests for all the usual properties (copyability, movability, construction) for all the relevant types I touched here. If it's helpful to single-step the commit history, I chunked the commits tiny and self-contained to make it easier. 

Also going to add @seelabs as a reviewer.

HowardHinnant commented on src/ripple/protocol/impl/STArray.cpp:
```diff
@@ -32,13 +32,13 @@ STArray::STArray()
 
-STArray::STArray (STArray&& other)
+STArray::STArray (STArray&& other) noexcept
     : STBase(other.getFName())
```
`STBase` is not nothrow constructible from `other.getFName()`, but I believe it could be made so.

JoeLoser replied:
Good point. Just made that possible in [this commit](https://github.com/ripple/rippled/pull/2551/commits/a4fc4d70bab422372568fb4248efab5318a9ece8). I appreciate the quick turn-around and excellent observations. I'll try to get to the other comments tonight or tomorrow.

HowardHinnant commented on src/ripple/protocol/impl/STArray.cpp:
```diff
@@ -32,13 +32,13 @@ STArray::STArray()
+STArray::STArray (STArray&& other) noexcept
     : STBase(other.getFName())
     , v_(std::move(other.v_))
 {
```
`CountedObject <STArray>` is not nothrow default constructible, but I believe it could be made so.

JoeLoser replied:
Fixed in 209c9fd7ad71ce26b868c1df276ff184f89f6660

HowardHinnant commented on src/ripple/protocol/impl/STArray.cpp:
```diff
@@ -32,13 +32,13 @@ STArray::STArray()
-STArray& STArray::operator= (STArray&& other)
+STArray& STArray::operator= (STArray&& other) noexcept
 {
     setFName(other.getFName());
```
`noexcept(setFName(other.getFName()))` evaluates to false, but I believe it could be made true.

JoeLoser replied:
Good catch. Done in 68c3ec29f

HowardHinnant commented on src/ripple/protocol/impl/STObject.cpp:
```diff
@@ -35,7 +35,7 @@ STObject::~STObject()
 
-STObject::STObject(STObject&& other)
+STObject::STObject(STObject&& other) noexcept
     : STBase(other.getFName())
```
`noexcept(other.getFName()` is false.

HowardHinnant replied:
The `STBase` constructor taking `other.getFName()` is not noexcept.

JoeLoser replied:
Another good catch. Fixed in f1e59f075

HowardHinnant commented on src/ripple/protocol/impl/STObject.cpp:
```diff
@@ -35,7 +35,7 @@ STObject::~STObject()
 }
 
-STObject::STObject(STObject&& other)
+STObject::STObject(STObject&& other) noexcept
```
`CountedObject<STObject>` is not noexcept default constructible.

JoeLoser replied:
Fixed in 209c9fd7ad71ce26b868c1df276ff184f89f6660

HowardHinnant commented on src/ripple/protocol/impl/STObject.cpp:
```diff
@@ -76,7 +76,7 @@ STObject::STObject (SerialIter& sit, SField const& name, int depth)
-STObject::operator= (STObject&& other)
+STObject::operator= (STObject&& other) noexcept
 {
     setFName(other.getFName());
```
`setFName(other.getFName())` is not noexcept.

JoeLoser replied:
Fixed in 68c3ec29f1be6088dba866e6685752cf40507b25

HowardHinnant commented on src/ripple/protocol/impl/STVar.cpp:
```diff
@@ -78,7 +78,7 @@ STVar::operator= (STVar const& rhs)
 STVar&
-STVar::operator= (STVar&& rhs)
+STVar::operator= (STVar&& rhs) noexcept
 {
```
`destroy()` is not noexcept.
`rhs.on_heap()` is not noexcept.
`p_ = rhs.p_->move(max_size, &d_)` is not noexcept.

HowardHinnant commented on src/ripple/protocol/impl/STVar.cpp:
```diff
@@ -49,7 +49,7 @@ STVar::STVar (STVar const& other)
-STVar::STVar (STVar&& other)
+STVar::STVar (STVar&& other) noexcept
 {
     if (other.on_heap())
```
`other.on_heap()` is not noexcept.
`p_ = other.p_->move(max_size, &d_)` is not noexcept.

HowardHinnant commented on src/ripple/protocol/impl/STVar.h:
```diff
@@ -58,11 +58,11 @@ class STVar
-    STVar (STBase&& t)
+    STVar (STBase&& t) noexcept
     {
         p_ = t.move(max_size, &d_);
```
`p_ = t.move(max_size, &d_)` is not noexcept.

JoeLoser commented:
So, I've addressed most of your comments with relevant fixes today. Thanks for the detailed review, @HowardHinnant!

The remaining issues are with `STArray` and `STObject` in making `CountedObject<STArray>` and `CountedSTObject` noexcept constructible. I took a first stab at it tonight and bypassed the [LWG issue](https://cplusplus.github.io/LWG/issue2116) that was getting me at first. Anyway, I'm still working on it.

HowardHinnant commented on src/ripple/protocol/impl/STBase.cpp:
```diff
@@ -51,6 +51,16 @@ STBase::operator= (const STBase& t)
+STBase&
+STBase::operator=(STBase&& other) noexcept
+{
+    if (!fName->isUseful())
```
`!fName->isUseful()` is not noexcept.

HowardHinnant commented on src/ripple/protocol/STBase.h:
```diff
@@ -157,7 +157,7 @@ class STBase
     static
     STBase*
-    emplace(std::size_t n, void* buf, T&& val)
+    emplace(std::size_t n, void* buf, T&& val) noexcept
```
`new U(std::forward<T>(val))` is not noexcept, and that one will be a little harder to fix.

JoeLoser replied:
I agree. It's gotten a bit out of scope than what I originally intended. I'm out of town  this weekend, but when I get back, I'll tidy up this PR to exclude the serialized types or create a new PR entirely for the non-ST types.

HowardHinnant commented on src/ripple/protocol/STBase.h:
```diff
@@ -157,7 +157,7 @@ class STBase
     static
     STBase*
-    emplace(std::size_t n, void* buf, T&& val)
+    emplace(std::size_t n, void* buf, T&& val) noexcept
```
`new(buf) U(std::forward<T>(val))` is not noexcept for `emplace<ripple::STBlob>`

JoeLoser replied:
Placing  
```
static_assert(noexcept(U(std::forward<T>(val))), "");
```
and letting it fire for all the contexts that make it fail made it easy to find and fix `STBlob`for which the copy-assignment operator is not `noexcept`. It appears that it could be made fine for most other protocol types. However, `STLedgerEntry` is where things go awry: when we look up a ledger entry type and do not find it, we throw `std::runtime_error`.  So the show may stop here with `noexcept`. I could mark `noexcept` conditionally for `emplace`, `move`, and the caller of `STBase::move`.

HowardHinnant replied:
The heap-based `new` is going to be a problem.  If the `STBase`-based classes are going to be done, lets put that work in its own PR.

seelabs replied:
I agree with Howard on this. The `ST*` classes are harder to get done. I'd rather we separated them out into their own PR as well.

HowardHinnant commented on src/ripple/protocol/impl/STBase.cpp:
```diff
@@ -36,7 +36,7 @@ STBase::STBase (SField const& n) noexcept
 
 STBase&
-STBase::operator= (const STBase& t)
+STBase::operator= (const STBase& t) noexcept
```
This `static_assert(noexcept(fName->isUseful()), "");` fires.

JoeLoser replied:
Right. Fixed in c3c66cba6

JoeLoser commented:
Going to close this PR for now and I'll open one next week with just the changes necessary for the non-ST-based types. Thanks for all the great feedback, @HowardHinnant.

HowardHinnant commented:
Thanks for the work!
